### PR TITLE
[Serverless Mini Agent] Use DogStatsD in Serverless

### PR DESCRIPTION
**Pull request by duncanpharvey**
# What does this PR do?

Starts a DogStatsD server when running the Serverless Mini Agent.

# Motivation

Allow runtime metrics and custom metrics to be sent from Azure Spring Apps and Azure Functions.

- https://datadoghq.atlassian.net/browse/SVLS-5177
- https://datadoghq.atlassian.net/browse/SVLS-5521

# Additional Notes

- Updated `main` method of Serverless to be asynchronous.
- Allow custom DogStatsD port with `DD_DOGSTATSD_PORT`
- Allow DogStatsD to be disabled with `DD_USE_DOGSTATSD`

# How to test the change?

* Build mini agent and deploy to Azure Spring App and Azure Function
* Confirm runtime metrics are sent successfully to Datadog

### CODE REVIEW

File: serverless/src/main.rs [hunk 1]
```diff
@@ -44,7 +67,61 @@ pub fn main() {
...
+    if !dd_use_dogstatsd {
+        return;
     }
+
+    let metrics_aggr = Arc::new(Mutex::new(
+        MetricsAggregator::new(Vec::new(), CONTEXTS).expect("Failed to create metrics aggregator"),
+    ));
+
+    info!("Starting DogStatsD");
+    let _ = start_dogstatsd(dd_dogstatsd_port, &metrics_aggr).await;
```
Comment (astuyve) [inline serverless/src/main.rs:86]: Probably okay for now not to use the cancel token but at some point you'll want to ensure you flush before the final shutdown of the container.

File: serverless/src/main.rs [hunk 2]
```diff
@@ -44,7 +67,61 @@ pub fn main() {
...
-        error!("Error when starting serverless trace mini agent: {e}");
+    tokio::spawn(async move {
+        let res = mini_agent.start_mini_agent().await;
+        if let Err(e) = res {
+            error!("Error when starting serverless trace mini agent: {e:?}");
+        }
+    });
+
+    if !dd_use_dogstatsd {
+        return;
```
Comment (astuyve) [inline serverless/src/main.rs:78]: I usually don't love an early return like this, I generally prefer if you'd just wrap the rest in this conditional just so somebody doesn't add more behavior after this line and not realize it isn't executing if statsd is disabled

**Review by astuyve — APPROVED** [review]
Couple notes, nothing major

File: serverless/src/main.rs (hunk 1 above)
Comment (duncanpharvey) [reply serverless/src/main.rs:86]: Yeah it works well enough for a beta just flushing every 10 seconds. But I still need to do some research for how to listen for shutdown events. Already have a card in our backlog for this one!

File: serverless/src/main.rs [hunk 3]
```diff
@@ -2,18 +2,41 @@
...
     Builder::from_env(env).target(Target::Stdout).init();
 
+    let dd_api_key: Option<String> = env::var("DD_API_KEY").ok();
+    let dd_dogstatsd_port: u16 = env::var("DD_DOGSTATSD_PORT")
+        .ok()
+        .and_then(|port| port.parse::<u16>().ok())
+        .unwrap_or(8125);
+    let dd_site = env::var("DD_SITE").unwrap_or_else(|_| "datadoghq.com".to_string());
+    let dd_use_dogstatsd = env::var("DD_USE_DOGSTATSD")
+        .map(|val| val != "false")
```
[the remark this answers is not in the document]
Comment (duncanpharvey) [reply serverless/src/main.rs]: Updated: https://github.com/DataDog/libdatadog/blob/01dcdf8002a611fc404e6c423ad1fc73f27dd514/serverless/src/main.rs#L36-L38

File: serverless/src/main.rs (hunk 2 above)
Comment (duncanpharvey) [reply serverless/src/main.rs]: Makes sense, I was trying to reduce the level of nesting but I agree it's more robust to not have the early return.

File: serverless/src/main.rs [hunk 4]
```diff
@@ -44,7 +67,76 @@ pub fn main() {
...
+        }
+    });
+
+    let mut metrics_flusher = if dd_use_dogstatsd {
+        let metrics_aggr = Arc::new(Mutex::new(
+            MetricsAggregator::new(Vec::new(), CONTEXTS)
+                .expect("Failed to create metrics aggregator"),
+        ));
+
+        info!("Starting dogstatsd");
```
Comment (duncanista) [inline serverless/src/main.rs:83]: Should this be `debug` logs instead of `info?`

File: serverless/src/main.rs [hunk 5]
```diff
@@ -44,7 +67,76 @@ pub fn main() {
...
+            }
+            None => {
+                error!("DD_API_KEY not set, won't flush metrics");
+                None
+            }
+        }
+    } else {
+        info!("dogstatsd disabled");
+        None
+    };
```
Comment (duncanista) [inline serverless/src/main.rs:104]: All of this could be abstracted into the `start_dogstatsd` method

File: serverless/src/main.rs [hunk 6]
```diff
@@ -44,7 +67,76 @@ pub fn main() {
...
+    flush_interval.tick().await; // discard first tick, which is instantaneous
+
+    loop {
+        flush_interval.tick().await;
+
+        if let Some(metrics_flusher) = metrics_flusher.as_mut() {
+            debug!("Flushing dogstatsd metrics");
+            metrics_flusher.flush().await;
+        } else {
+            debug!("dogstatsd disabled, not flushing metrics");
```
Comment (duncanista) [inline serverless/src/main.rs:116]: You don't are already logging this, might not be worth having this

File: trace-mini-agent/src/config.rs [hunk 7]
```diff
@@ -41,6 +42,10 @@ impl Config {
             anyhow::anyhow!("Unable to identify environment. Shutting down Mini Agent.")
         })?;
 
+        let dd_dogstatsd_port: u16 = env::var("DD_DOGSTATSD_PORT")
+            .ok()
+            .and_then(|port| port.parse::<u16>().ok())
+            .unwrap_or(8125);
```
Comment (duncanista) [inline trace-mini-agent/src/config.rs:48]: magic number

File: serverless/src/main.rs (hunk 5 above)
Comment (duncanpharvey) [reply serverless/src/main.rs]: Refactored to move the metrics aggregator and flusher into `start_dogstatsd`

File: serverless/src/main.rs (hunk 4 above)
Comment (duncanpharvey) [reply serverless/src/main.rs]: Changed this and the `Starting serverless trace mini agent` log to both be `debug` logs.

File: trace-mini-agent/src/config.rs (hunk 7 above)
Comment (duncanpharvey) [reply trace-mini-agent/src/config.rs:48]: Added an environment variable `DEFAULT_DOGSTATSD_PORT` for this as well as an `AGENT_HOST` environment variable.

File: serverless/src/main.rs (hunk 5 above)
Comment (duncanpharvey) [reply serverless/src/main.rs]: https://github.com/DataDog/libdatadog/blob/55201cf0181104a17e7eefc13baa7804ab1de595/serverless/src/main.rs#L102-L143

**Pull request merged**